usr/share/rear/layout/save/default/600_snapshot_files.sh

```diff
@@ -10,17 +10,40 @@
 # we risk inconsistencies between the restored files and recreated layout,
 # or failures of UUID migration.
 
-local file obj
+local file obj final_file symlink_target
 
 # The patterns are relative to /, change directory there
 # so that the shell finds the files during pathname expansion
 pushd / >/dev/null
 # The variable expansion is deliberately not quoted in order to perform
 # pathname expansion on the variable value.
 for file in $FILES_TO_PATCH_PATTERNS ; do
-    if ! IsInArray "/$file" "${CHECK_CONFIG_FILES[@]}" ; then
-        CHECK_CONFIG_FILES+=( "/$file" )
+    final_file="/$file"
+    # Symlink handling: replace symlinks by targets (not strictly necessary, but consistent
+    # with 280_migrate_uuid_tags.sh), avoid dead symlinks, and symlinks to files
+    # on dynamic filesystem ( /proc etc.) - they are expected to change
+    # and validating their checksums has no sense
+    if ! IsInArray "$final_file" "${CHECK_CONFIG_FILES[@]}" ; then
```

Comment:
Only a cosmetic code simplification proposal:

Wouldn't it be simpler to do the check
if a file from the FILES_TO_PATCH_PATTERNS evaluation
is already in CHECK_CONFIG_FILES
first of all in the for loop like
```
for file in $FILES_TO_PATCH_PATTERNS ; do
    IsInArray "$file" "${CHECK_CONFIG_FILES[@]}" && continue
    final_file="/$file"
    # Symlink handling: ...
        ...
                   IsInArray "$symlink_target" "${CHECK_CONFIG_FILES[@]}" && continue     
```

Reply:
One can't do `IsInArray "$file"`, because `$file` is relative and `CHECK_CONFIG_FILES` are absolute.
Regarding this code in general: I now believe that copying the symlink handling entirely from `280_migrate_uuid_tags.sh` is a bad idea. We should calculate the hash of the symlink and not attempt resolving the symlink to its target (except for checking of broken symlinks). The reason is, what if the symlink changes to another target? If we check the hash of the target only, we see no change, but in practice this is a config file change.

Reply:
I have seen my relative `$file` mistake and already fixed my code proposal.

I agree that we should also check where a symlink points to
but I think only checking where a symlink points to is insufficient
because I think we must also check the symlink target content.

When we only check where a symlink points to
but not also automatically check symlink target contents
we would miss symlink target content changes
when not all symlink targets are explicitly specified in
CHECK_CONFIG_FILES and FILES_TO_PATCH_PATTERNS.

E.g. a user may specify some /path/to/this in
CHECK_CONFIG_FILES or FILES_TO_PATCH_PATTERNS
but /path/to/this is a symlink that points to /path/to/that and
then we should ensure /path/to/that gets also automatically checked.

Reply:
Oh, I want to check the contents, that's sure, but by supplying the symlink itself to `md5sum`, not by resolving it to the target and supplying the target. (This affects what will get recorded in the checksum file, and this what will be checked when the time to verify the checksums comes.)

Reply:
Symlinks always confuse me.

As far as I see `md5sum symlink` does not calculate the hash of the symlink
but resolves the symlink to its target and calculate the hash of its target
and I don't see "man md5sum" mention symlinks:
```
# echo content >file1

# echo content >file2

# ln -s file1 symlink

# file symlink
symlink: symbolic link to file1

# md5sum symlink
f75b8179e4bbe7e2b4a074dcef62de95  symlink

# ln -sf file2 symlink

# file symlink
symlink: symbolic link to file2

# md5sum symlink
f75b8179e4bbe7e2b4a074dcef62de95  symlink
```

Reply:
I didn't notice your
https://github.com/rear/rear/pull/2795#discussion_r862718197
while I wrote my
https://github.com/rear/rear/pull/2795#discussion_r862721784

Reply:
I think I understand now what you intend to check:

When there is a change where a symlink points to
and the new symlink target content is different
(the usual case in practice)
then "md5sum" will correctly detect it.

When there is a change where a symlink points to
and the new symlink target content is the same
(likely this does not happen in practice)
then "md5sum" will also correctly show all is OK.

Reply:
> When there is a change where a symlink points to and the new symlink target content is different (the usual case in practice) then "md5sum" will correctly detect it.
> 
> When there is a change where a symlink points to and the new symlink target content is the same (likely this does not happen in practice) then "md5sum" will also correctly show all is OK.

Yes, and the first case is not detected properly now, that's why I want to change the symlink handling.

By the way, your "cosmetic code simplification" is also a bugfix: I realized that the file was added to `CHECK_CONFIG_FILES` despite being already there, and your proposal fixes it.

Reply:
For the fun of it:

With symlinks to symlinks
there can be a change where a symlink points to
but the new symlink target content is still the same:
```
# echo content >target

# ln -s target symlink1

# ln -s target symlink2

# ln -s symlink1 symlink

# file symlink
symlink:  symbolic link to symlink1

# md5sum symlink
f75b8179e4bbe7e2b4a074dcef62de95  symlink

# ln -sf symlink2 symlink

# file symlink
symlink: symbolic link to symlink2

# md5sum symlink
f75b8179e4bbe7e2b4a074dcef62de95  symlink
```

RFC 1925 item 6a:
"It is always possible to add another level of indirection."
:-)

Reply:
This is not a problem for my approach as I don't care what is the actual symlink target at all, only what its content is.

Reply:
I didn't mean my
https://github.com/rear/rear/pull/2795#discussion_r862735745
as a counter example against the intended check
because both cases in
https://github.com/rear/rear/pull/2795#discussion_r862729011
do work correctly.

I meant it only as an example for the second case in
https://github.com/rear/rear/pull/2795#discussion_r862729011
that might even happen in practice somewhere "out there".